samcli/commands/local/lib/sam_api_provider.py

```diff
@@ -197,6 +204,40 @@ def _merge_apis(collector):
 
         return list(result)
 
+    @staticmethod
+    def _normalize_api_stage(apis):
+        """
+        Explicit APIs are required to have a stage name.
+        Functions with events that reference an explicit API need to use the same stage name as the API
+
+        Parameters
+        ----------
+        apis : list of samcli.commands.local.lib.provider.Api
+            list of APIs that should have the same stage name
+
+        Returns
+        -------
+        list of samcli.commands.local.lib.provider.Api
+            List of normalized APIs
+
+        """
+
+        result = list()
+        stage_name = None
+
+        # Find the stage name
+        for api in apis:
```

Comment:
If I remember right, it's not that straightforward (yet). When SAM CLI is parsing the template, it can correctly find the `StageName` for an API (since it is a required parameter) but not for a lambda event (not a required parameter). Events are sometimes associated with an API, so it could inherit a stage name (which is what this is searching for)- or it could not have an associated API, at which point SAM creates a default one with default stage values. We don't know which one it is associated with until they are all grouped together in this `apis` list that I iterate over here.